addon/services/i18n.js

```diff
@@ -19,11 +19,25 @@ export default Parent.extend(Ember.Evented, {
   locales: computed(getLocales),
 
   // @public
+  // Whether to allow overriding the locale for individual translations.
+  allowLocaleOverride: undefined,
+
+  // @public
   //
   // Returns the translation `key` interpolated with `data`
   // in the current `locale`.
   t: function(key, data = {}) {
-    const locale = this.get('_locale');
+    let locale = this.get('_locale');
+    if (data.locale && this.get('allowLocaleOverride') !== false) {
+      Ember.warn(
+        'Specifying an interpolation key named `locale` now overrides the target locale for translation. To silence this warning, set ENV.i18n.allowLocaleOverride to true. To disable this behavior and treat `locale` as a normal interpolation key, set ENV.i18n.allowLocaleOverride to false.',
+        (this.get('allowLocaleOverride') !== undefined),
+        { id: 'ember-i18n.locale-override' }
+      );
+      locale = new Locale(data.locale, this.container);
```

Comment:
1. we only want to warn if `data.locale` is defined
 2. we can make it slightly more efficient by saving off the warning text as a `const` at the top of the file
 3. we could cache `Locale`s in a closure-local map in the `locale.js` file:

```js
const seenLocales = {};

function Locale(name, container) {
  if (seenLocales[name]) { return seenLocales[name]; }
  seenLocales[name] = this;
  ...
}
```